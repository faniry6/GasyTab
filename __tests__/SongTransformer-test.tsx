import 'react-native';
import React from 'react';
import { Text } from 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import SongTransformer, { transposeSong, getChords } from '../app/components/SongTransformer';
import ChordSheetJS from 'chordsheetjs'
import Chord from 'chordjs'

it('renders correctly', () => {
  let chordProSong = `
{title: Let it be}
{artist: The Beatles}
{soc}
Let it [Am]be, let it [C/G]be, let it [F]be, let it [C]be big[Cm]word
Let it [Am]be, let it [C/G]be, let it [F]be, let it [C]be
{eoc}

{comment: Intro [C] [D]}
Let it [Am]be, let it [C/G]be, let it [F]be, let it [C]be
[C]Whisper words of w[G]isdom, let it [F]be [C/E]      [Dm]     [C] 
{sot}
C|---
E|---
{eot}
`
  const tree = renderer.create(
    <SongTransformer
      chordProSong={chordProSong}
      transposeDelta={0}
      children={(songProps) => <Text>{songProps.htmlSong}</Text>} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('hide tabs', () => {
  let chordProSong = `
{title: Let it be}
{artist: The Beatles}
{soc}
Let it [Am]be, let it [C/G]be, let it [F]be, let it [C]be big[Cm]word
Let it [Am]be, let it [C/G]be, let it [F]be, let it [C]be
{eoc}

{comment: Intro [C] [D]}
Let it [Am]be, let it [C/G]be, let it [F]be, let it [C]be
{sot}
C|---
E|---
{eot}
[C]Whisper words of w[G]isdom, let it [F]be [C/E]      [Dm]     [C] 
{sot}
C|---
E|---
{eot}`
  const tree = renderer.create(
    <SongTransformer
      chordProSong={chordProSong}
      transposeDelta={0}
      showTabs={false}
      children={(songProps) => <Text>{songProps.htmlSong}</Text>} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('hide tabs with different line breaking formats', () => {
  let chordProSong = `
{sot}E|---{eot}
{sot}
C|---
E|---{eot}
{sot}
C|---
E|---
{eot}{sot}
C|---
E|---
{eot}`
  const tree = renderer.create(
    <SongTransformer
      chordProSong={chordProSong}
      transposeDelta={0}
      showTabs={false}
      children={(songProps) => <Text>{songProps.htmlSong}</Text>} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('transpose comment tag correctly', () => {
  let chordProSong =
    "{comment: Intro [C] [D7]  [F#]}\n" +
    "More lyrics [C] [D7]  [F#]"
  let song = new ChordSheetJS.ChordProParser().parse(chordProSong)
  let transposedSong = transposeSong(song, 1)
  let chordProTransposed = new ChordSheetJS.ChordProFormatter().format(transposedSong)
  let res =
    "{comment: Intro [Db] [Eb7]  [G]}\n" +
    "More lyrics [Db] [Eb7]  [G]"
  expect(chordProTransposed).toBe(res)
})

it('return all chords without repeating', () => {
  let chordProSong =
    "{comment: Intro [E] [G7]  [F#]}\n" +
    "More lyrics [C] [D7]  [F#]"
  let song = new ChordSheetJS.ChordProParser().parse(chordProSong)
  let allChords = getChords(song)
  let chords: Chord[] = [
    Chord.parse("E")!,
    Chord.parse("G7")!,
    Chord.parse("F#")!,
    Chord.parse("C")!,
    Chord.parse("D7")!,
  ]
  allChords.forEach((chord, index) => {
    expect(chord.toString()).toBe(chords[index].toString())
  })
})