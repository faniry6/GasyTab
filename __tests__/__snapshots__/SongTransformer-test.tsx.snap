// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hide tabs 1`] = `
<Text>
  &lt;p class="line"&gt;&lt;span class="title"&gt;Let it be&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;span class="artist"&gt;The Beatles&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
&lt;p class="line"&gt;Let it &lt;span class="chord"&gt;Am&lt;/span&gt;be, let it &lt;span class="chord"&gt;C/G&lt;/span&gt;be, let it &lt;span class="chord"&gt;F&lt;/span&gt;be, let it &lt;span class="chord"&gt;C&lt;/span&gt;be &lt;span class="word"&gt;big&lt;span class="chord"&gt;Cm&lt;/span&gt;word&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;Let it &lt;span class="chord"&gt;Am&lt;/span&gt;be, let it &lt;span class="chord"&gt;C/G&lt;/span&gt;be, let it &lt;span class="chord"&gt;F&lt;/span&gt;be, let it &lt;span class="chord"&gt;C&lt;/span&gt;be&lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;span class="comment"&gt;Intro &lt;span class="chord chord-inline"&gt;C&lt;/span&gt; &lt;span class="chord chord-inline"&gt;D&lt;/span&gt;&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;Let it &lt;span class="chord"&gt;Am&lt;/span&gt;be, let it &lt;span class="chord"&gt;C/G&lt;/span&gt;be, let it &lt;span class="chord"&gt;F&lt;/span&gt;be, let it &lt;span class="chord"&gt;C&lt;/span&gt;be&lt;/p&gt;
&lt;p class="line"&gt;&lt;span class="chord"&gt;C&lt;/span&gt;Whisper words of &lt;span class="word"&gt;w&lt;span class="chord"&gt;G&lt;/span&gt;isdom&lt;/span&gt;, let it &lt;span class="chord"&gt;F&lt;/span&gt;be &lt;span class="chord"&gt;C/E&lt;/span&gt;      &lt;span class="chord"&gt;Dm&lt;/span&gt;     &lt;span class="chord"&gt;C&lt;/span&gt;  &lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
</Text>
`;

exports[`hide tabs with different line breaking formats 1`] = `
<Text>
  &lt;p class="line"&gt;&lt;/p&gt;
</Text>
`;

exports[`renders correctly 1`] = `
<Text>
  &lt;p class="line"&gt;&lt;span class="title"&gt;Let it be&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;span class="artist"&gt;The Beatles&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
&lt;p class="line"&gt;Let it &lt;span class="chord"&gt;Am&lt;/span&gt;be, let it &lt;span class="chord"&gt;C/G&lt;/span&gt;be, let it &lt;span class="chord"&gt;F&lt;/span&gt;be, let it &lt;span class="chord"&gt;C&lt;/span&gt;be &lt;span class="word"&gt;big&lt;span class="chord"&gt;Cm&lt;/span&gt;word&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;Let it &lt;span class="chord"&gt;Am&lt;/span&gt;be, let it &lt;span class="chord"&gt;C/G&lt;/span&gt;be, let it &lt;span class="chord"&gt;F&lt;/span&gt;be, let it &lt;span class="chord"&gt;C&lt;/span&gt;be&lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;/p&gt;
&lt;p class="line"&gt;&lt;span class="comment"&gt;Intro &lt;span class="chord chord-inline"&gt;C&lt;/span&gt; &lt;span class="chord chord-inline"&gt;D&lt;/span&gt;&lt;/span&gt;&lt;/p&gt;
&lt;p class="line"&gt;Let it &lt;span class="chord"&gt;Am&lt;/span&gt;be, let it &lt;span class="chord"&gt;C/G&lt;/span&gt;be, let it &lt;span class="chord"&gt;F&lt;/span&gt;be, let it &lt;span class="chord"&gt;C&lt;/span&gt;be&lt;/p&gt;
&lt;p class="line"&gt;&lt;span class="chord"&gt;C&lt;/span&gt;Whisper words of &lt;span class="word"&gt;w&lt;span class="chord"&gt;G&lt;/span&gt;isdom&lt;/span&gt;, let it &lt;span class="chord"&gt;F&lt;/span&gt;be &lt;span class="chord"&gt;C/E&lt;/span&gt;      &lt;span class="chord"&gt;Dm&lt;/span&gt;     &lt;span class="chord"&gt;C&lt;/span&gt;  &lt;/p&gt;
&lt;div class="tab"&gt;&lt;div class="tab-line"&gt;CE&lt;/div&gt;
&lt;div class="tab-line"&gt;||&lt;/div&gt;
&lt;div class="tab-line"&gt;--&lt;/div&gt;
&lt;div class="tab-line"&gt;--&lt;/div&gt;
&lt;div class="tab-line"&gt;--&lt;/div&gt;
&lt;/div&gt;
&lt;p class="line"&gt;&lt;/p&gt;
</Text>
`;
